File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/TableManagerSamples/RemoveOrder/src/TableListener.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using System.Threading;
using fxcore2;

namespace RemoveOrder
{
    class TableListener : IO2GTableListener
    {
        private ResponseListener mResponseListener;
        private string mRequestID;
        private string mOrderID;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="responseListener"></param>
        public TableListener(ResponseListener responseListener)
        {
            mResponseListener = responseListener;
            mRequestID = string.Empty;
            mOrderID = string.Empty;
        }

        public void SetRequestID(string sRequestID)
        {
            mRequestID = sRequestID;
        }

        public string GetOrderID()
        {
            return mOrderID;
        }

        #region IO2GTableListener Members

        // Implementation of IO2GTableListener interface public method onAdded
        public void onAdded(string sRowID, O2GRow rowData)
        {
            if (rowData.TableType == O2GTableType.Orders)
            {
                O2GOrderRow orderRow = (O2GOrderRow)rowData;
                if (mRequestID.Equals(orderRow.RequestID))
                {
                    if (IsLimitEntryOrder(orderRow) && string.IsNullOrEmpty(mOrderID))
                    {
                        mOrderID = orderRow.OrderID;
                        Console.WriteLine("The order has been added. Order ID: {0}, Rate: {1}, Time In Force: {2}",
                            orderRow.OrderID,
                            orderRow.Rate,
                            orderRow.TimeInForce);
                        mResponseListener.StopWaiting();
                    }
                }
            }
        }

        // Implementation of IO2GTableListener interface public method onChanged
        public void onChanged(string sRowID, O2GRow rowData)
        {
        }

        // Implementation of IO2GTableListener interface public method onDeleted
        public void onDeleted(string sRowID, O2GRow rowData)
        {
            if (rowData.TableType == O2GTableType.Orders)
            {
                O2GOrderRow orderRow = (O2GOrderRow)rowData;
                if (mRequestID.Equals(orderRow.RequestID))
                {
                    if (!string.IsNullOrEmpty(mOrderID))
                    {
                        Console.WriteLine("The order has been deleted. Order ID: {0}", orderRow.OrderID);
                        mResponseListener.StopWaiting();
                    }
                }
            }
        }

        public void onStatusChanged(O2GTableStatus status)
        {
        }

        #endregion

        public void SubscribeEvents(O2GTableManager manager)
        {
            O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.Orders);
            ordersTable.subscribeUpdate(O2GTableUpdateType.Insert, this);
            ordersTable.subscribeUpdate(O2GTableUpdateType.Delete, this);
        }

        public void UnsubscribeEvents(O2GTableManager manager)
        {
            O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.Orders);
            ordersTable.unsubscribeUpdate(O2GTableUpdateType.Insert, this);
            ordersTable.unsubscribeUpdate(O2GTableUpdateType.Delete, this);
        }

        private bool IsLimitEntryOrder(O2GOrderRow order)
        {
            return order.Type.StartsWith("LE");
        }

    }
}

File: samples/macOS/java/TableManagerSamples/RemoveOrder/src/TableListener.java
package removeorder;

import com.fxcore2.*;

public class TableListener implements IO2GTableListener {
    private ResponseListener mResponseListener;
    private String mRequestID;
    private String mOrderID;

    // ctor
    public TableListener(ResponseListener responseListener) {
        mResponseListener = responseListener;
        mRequestID = "";
        mOrderID = "";
    }

    public void setRequestID(String sRequestID) {
        mRequestID = sRequestID;
    }

    public String getOrderID() {
        return mOrderID;
    }

    // Implementation of IO2GTableListener interface public method onAdded
    public void onAdded(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.ORDERS) {
            O2GOrderRow orderRow = (O2GOrderRow)rowData;
            if (mRequestID.equals(orderRow.getRequestID())) {
                if (isLimitEntryOrder(orderRow) && mOrderID.isEmpty()) {
                    mOrderID = orderRow.getOrderID();
                    System.out.println(String.format("The order has been added. OrderID=%s, Type=%s, BuySell=%s, Rate=%s, TimeInForce=%s",
                            orderRow.getOrderID(), orderRow.getType(), orderRow.getBuySell(), orderRow.getRate(), orderRow.getTimeInForce()));
                    mResponseListener.stopWaiting();
                }
            }
        }
    }

    // Implementation of IO2GTableListener interface public method onChanged
    public void onChanged(String sRowID, O2GRow rowData) {
    }

    // Implementation of IO2GTableListener interface public method onDeleted
    public void onDeleted(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.ORDERS) {
            O2GOrderRow orderRow = (O2GOrderRow)rowData;
            if (mRequestID.equals(orderRow.getRequestID())) {
                if (!mOrderID.isEmpty()) {
                    System.out.println(String.format("The order has been deleted. Order ID: %s", orderRow.getOrderID()));
                    mResponseListener.stopWaiting();
                }
            }
        }
    }

    public void onStatusChanged(O2GTableStatus status) {
    }

    public void subscribeEvents(O2GTableManager manager) {
        O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.ORDERS);
        ordersTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        ordersTable.subscribeUpdate(O2GTableUpdateType.DELETE, this);
    }

    public void unsubscribeEvents(O2GTableManager manager) {
        O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.ORDERS);
        ordersTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        ordersTable.unsubscribeUpdate(O2GTableUpdateType.DELETE, this);
    }

    private boolean isLimitEntryOrder(O2GOrderRow order)
    {
        return order.getType().startsWith("LE");
    }
}
